# I. Introduction

One of the most controversial issues in the U.S. educational system is the efficacy of standardized tests, and whether they're unfair to certain groups. Given our prior knowledge of this topic, investigating the correlations between SAT scores and demographics might be an interesting angle to take. We could correlate SAT scores with factors like race, gender, income, and more.

The SAT, or Scholastic Aptitude Test, is an exam that U.S. high school students take before applying to college. Colleges take the test scores into account when deciding who to admit, so it's fairly important to perform well on it.

The test consists of three sections, each of which has 800 possible points. The combined score is out of 2,400 possible points (while this number has changed a few times, the data set for our project is based on 2,400 total points). Organizations often rank high schools by their average SAT scores. The scores are also considered a measure of overall school district quality.

New York City makes its data on high school SAT scores available online, as well as the demographics for each high school. We'll need to combine several interrelated datasets into one before we can find correlations. Below are the datasets we'll be using.
 - SAT scores by school - SAT scores for each high school in New York City
 - School attendance - Attendance information for each school in New York City
 - Class size - Information on class size for each school
 - AP test results - Advanced Placement (AP) exam results for each high school
 - Graduation outcomes - The percentage of students who graduated, and other outcome information
 - Demographics - Demographic information for each school
 - School survey - Surveys of parents, teachers, and students at each school

# II. Prepping the Data Sets

In [1]:
import pandas as pd

data_files = [
    "Graduation.csv",
    "Demographics.csv",
    "AP_Results.csv",
    "Class_Size.csv",
    "SAT_Results.csv",
    "HS_Directory.csv"
]

data = {}
for file in data_files:
    df = pd.read_csv('Schools/{0}'.format(file))
    key_name = file.replace(".csv", "")
    data[key_name] = df

### High School Directory

In [2]:
data['HS_Directory'].head(3)

dbn                                  school_name   borough  \
0  17K548          Brooklyn School for Music & Theatre  Brooklyn   
1  09X543             High School for Violin and Dance     Bronx   
2  09X327  Comprehensive Model School Project M.S. 327     Bronx   

  building_code  phone_number    fax_number  grade_span_min  grade_span_max  \
0          K440  718-230-6250  718-230-6262             9.0              12   
1          X400  718-842-0687  718-589-9849             9.0              12   
2          X240  718-294-8111  718-294-8109             6.0              12   

   expgrade_span_min  expgrade_span_max  \
0                NaN                NaN   
1                NaN                NaN   
2                NaN                NaN   

                         ...                         priority08 priority09  \
0                        ...                                NaN        NaN   
1                        ...                                NaN        NaN   
2                        ...                                NaN        NaN   

  priority10                                         Location 1  \
0        NaN  883 Classon Avenue\nBrooklyn, NY 11225\n(40.67...   
1        NaN  1110 Boston Road\nBronx, NY 10456\n(40.8276026...   
2        NaN  1501 Jerome Avenue\nBronx, NY 10452\n(40.84241...   

  Community Board  Council District Census Tract        BIN           BBL  \
0             9.0              35.0        213.0  3029686.0  3.011870e+09   
1             3.0              16.0        135.0  2004526.0  2.026340e+09   
2             4.0              14.0        209.0  2008336.0  2.028590e+09   

                                                 NTA  
0  Crown Heights South                           ...  
1  Morrisania-Melrose                            ...  
2  West Concourse                                ...  

[3 rows x 64 columns]

### Graduation Outcomes

In [3]:
data['Graduation'].head(3)

Demographic     DBN                            School Name Cohort  \
0  Total Cohort  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL   2003   
1  Total Cohort  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL   2004   
2  Total Cohort  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL   2005   

   Total Cohort Total Grads - n Total Grads - % of cohort Total Regents - n  \
0             5               s                         s                 s   
1            55              37                     67.3%                17   
2            64              43                     67.2%                27   

  Total Regents - % of cohort Total Regents - % of grads  \
0                           s                          s   
1                       30.9%                      45.9%   
2                       42.2%                      62.8%   

             ...            Regents w/o Advanced - n  \
0            ...                                   s   
1            ...                                  17   
2            ...                                  27   

  Regents w/o Advanced - % of cohort Regents w/o Advanced - % of grads  \
0                                  s                                 s   
1                              30.9%                             45.9%   
2                              42.2%                             62.8%   

  Local - n Local - % of cohort   Local - % of grads Still Enrolled - n  \
0         s                   s                    s                  s   
1        20               36.4%                54.1%                 15   
2        16                 25%  37.200000000000003%                  9   

  Still Enrolled - % of cohort Dropped Out - n Dropped Out - % of cohort  
0                            s               s                         s  
1                        27.3%               3                      5.5%  
2                        14.1%               9                     14.1%  

[3 rows x 23 columns]

### Class Size

In [4]:
data['Class_Size'].head(3)

CSD BOROUGH SCHOOL CODE                SCHOOL NAME GRADE  PROGRAM TYPE  \
0    1       M        M015  P.S. 015 Roberto Clemente     0K       GEN ED   
1    1       M        M015  P.S. 015 Roberto Clemente     0K          CTT   
2    1       M        M015  P.S. 015 Roberto Clemente     01       GEN ED   

  CORE SUBJECT (MS CORE and 9-12 ONLY) CORE COURSE (MS CORE and 9-12 ONLY)  \
0                                    -                                   -   
1                                    -                                   -   
2                                    -                                   -   

  SERVICE CATEGORY(K-9* ONLY)  NUMBER OF STUDENTS / SEATS FILLED  \
0                           -                               19.0   
1                           -                               21.0   
2                           -                               17.0   

   NUMBER OF SECTIONS  AVERAGE CLASS SIZE  SIZE OF SMALLEST CLASS  \
0                 1.0                19.0                    19.0   
1                 1.0                21.0                    21.0   
2                 1.0                17.0                    17.0   

   SIZE OF LARGEST CLASS DATA SOURCE  SCHOOLWIDE PUPIL-TEACHER RATIO  
0                   19.0         ATS                             NaN  
1                   21.0         ATS                             NaN  
2                   17.0         ATS                             NaN

### Demographics

In [5]:
data['Demographics'].head(3)

DBN                       Name  schoolyear fl_percent  frl_percent  \
0  01M015  P.S. 015 ROBERTO CLEMENTE    20052006       89.4          NaN   
1  01M015  P.S. 015 ROBERTO CLEMENTE    20062007       89.4          NaN   
2  01M015  P.S. 015 ROBERTO CLEMENTE    20072008       89.4          NaN   

   total_enrollment prek   k grade1 grade2    ...     black_num black_per  \
0               281   15  36     40     33    ...            74      26.3   
1               243   15  29     39     38    ...            68      28.0   
2               261   18  43     39     36    ...            77      29.5   

  hispanic_num hispanic_per white_num white_per male_num male_per female_num  \
0          189         67.3         5       1.8    158.0     56.2      123.0   
1          153         63.0         4       1.6    140.0     57.6      103.0   
2          157         60.2         7       2.7    143.0     54.8      118.0   

  female_per  
0       43.8  
1       42.4  
2       45.2  

[3 rows x 38 columns]

### AP Results

In [6]:
data['AP_Results'].head(3)

DBN                    SchoolName AP Test Takers  Total Exams Taken  \
0  01M448  UNIVERSITY NEIGHBORHOOD H.S.              39                49   
1  01M450        EAST SIDE COMMUNITY HS              19                21   
2  01M515           LOWER EASTSIDE PREP              24                26   

  Number of Exams with scores 3 4 or 5  
0                                   10  
1                                    s  
2                                   24

### SAT Results

In [7]:
data['SAT_Results'].head(3)

DBN                                    SCHOOL NAME  \
0  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL STUDIES   
1  01M448            UNIVERSITY NEIGHBORHOOD HIGH SCHOOL   
2  01M450                     EAST SIDE COMMUNITY SCHOOL   

  Num of SAT Test Takers SAT Critical Reading Avg. Score SAT Math Avg. Score  \
0                     29                             355                 404   
1                     91                             383                 423   
2                     70                             377                 402   

  SAT Writing Avg. Score  
0                    363  
1                    366  
2                    370

### School Survey
 
Our survey data, survey_all.txt and survey_d75.txt, are in a slightly more complex format. They are tab delimited and encoded with the rarely used Windows-1252 encoding. An encoding defines how a computer stores the contents of a file in binary. The most common encodings are UTF-8 and ASCII. Windows-1252 can cause errors read in without specifying the encoding. After the survey data is properly read in, the two files will be combined into a single dataframe.

In [8]:
all_survey = pd.read_csv("Schools/survey_all.txt", delimiter='\t', encoding='windows-1252')
d75_survey = pd.read_csv('Schools/survey_d75.txt', delimiter='\t', encoding='windows-1252')
survey = pd.concat([all_survey, d75_survey], axis=0)
survey.head(3)

N_p  N_s   N_t  aca_p_11  aca_s_11  aca_t_11  aca_tot_11    bn  com_p_11  \
0   90.0  NaN  22.0       7.8       NaN       7.9         7.9  M015       7.6   
1  161.0  NaN  34.0       7.8       NaN       9.1         8.4  M019       7.6   
2  367.0  NaN  42.0       8.6       NaN       7.5         8.0  M020       8.3   

   com_s_11   ...    t_q8c_1  t_q8c_2  t_q8c_3 t_q8c_4  t_q9  t_q9_1  t_q9_2  \
0       NaN   ...       29.0     67.0      5.0     0.0   NaN     5.0    14.0   
1       NaN   ...       74.0     21.0      6.0     0.0   NaN     3.0     6.0   
2       NaN   ...       33.0     35.0     20.0    13.0   NaN     3.0     5.0   

   t_q9_3  t_q9_4  t_q9_5  
0    52.0    24.0     5.0  
1     3.0    78.0     9.0  
2    16.0    70.0     5.0  

[3 rows x 2773 columns]

There are over 2000 columns. We need to filter and remove the columns we don't need to better find correlations within the data. We have a data dictionary for the survey data that tells us what each column represents so we may determine which columns to use.

Per the dictionary, we'll want to use the following columns ["dbn", "rr_s", "rr_t", "rr_p", "N_s", "N_t", "N_p", "saf_p_11", "com_p_11", "eng_p_11", "aca_p_11", "saf_t_11", "com_t_11", "eng_t_11", "aca_t_11", "saf_s_11", "com_s_11", "eng_s_11", "aca_s_11", "saf_tot_11", "com_tot_11", "eng_tot_11", "aca_tot_11"]. These columns will tell us about how parents, teachers, and students feel about school safety, academic performance, and more. 

The survey data has a dbn column, a unique identification code for each school. The column name needs to be converted to uppercase (DBN) to make it consistent with the other data sets and avoid any errors when combining them.

In [9]:
survey['DBN'] = survey['dbn']
columns = ['DBN', "rr_s", "rr_t", "rr_p", "N_s", "N_t", "N_p", "saf_p_11", "com_p_11", "eng_p_11", "aca_p_11", "saf_t_11", "com_t_11", "eng_t_11", "aca_t_11", "saf_s_11", "com_s_11", "eng_s_11", "aca_s_11", "saf_tot_11", "com_tot_11", "eng_tot_11", "aca_tot_11"]
survey = survey.loc[:,columns]
data['survey'] = survey
survey.head(3)

DBN  rr_s  rr_t  rr_p  N_s   N_t    N_p  saf_p_11  com_p_11  eng_p_11  \
0  01M015   NaN    88    60  NaN  22.0   90.0       8.5       7.6       7.5   
1  01M019   NaN   100    60  NaN  34.0  161.0       8.4       7.6       7.6   
2  01M020   NaN    88    73  NaN  42.0  367.0       8.9       8.3       8.3   

      ...      eng_t_11  aca_t_11  saf_s_11  com_s_11  eng_s_11  aca_s_11  \
0     ...           7.6       7.9       NaN       NaN       NaN       NaN   
1     ...           8.9       9.1       NaN       NaN       NaN       NaN   
2     ...           6.8       7.5       NaN       NaN       NaN       NaN   

   saf_tot_11  com_tot_11  eng_tot_11  aca_tot_11  
0         8.0         7.7         7.5         7.9  
1         8.5         8.1         8.2         8.4  
2         8.2         7.3         7.5         8.0  

[3 rows x 23 columns]

### Observations

- Each data set appears to either have a DBN column, or the information we need to create one. That means we can use a DBN column to combine the data sets. First we'll pinpoint matching rows from different data sets by looking for identical DBNs, then group all of their columns together in a single data set.


- Some fields look interesting for mapping, Location 1 in particular, which contains coordinates inside a larger string.


- Some of the data sets appear to contain multiple rows for each school (because the rows have duplicate DBN values). That means we’ll have to do some preprocessing to ensure that each DBN is unique within each data set. If we don't do this, we'll run into problems when we combine the data sets, because we might be merging two rows in one data set with one row in another data set.

### Standardizing "DBN" Fields

The class size and high school directory data sets do not have a DBN column. The directory's "dbn" column needs to be renamed to "DBN" and the class size data doesn't have the column at all.

If we compare the class size and SAT results data sets, we can tell that the DBN in the SAT results data is just a combination of the "CSD" and "SCHOOL CODE" columns in the class size data. The main difference is that the DBN is padded, so that the CSD portion of it always consists of two digits. That means we'll need to add a leading 0 to the CSD if it is less than two digits long.

In [10]:
data["HS_Directory"]["DBN"] = data["HS_Directory"]["dbn"]

def padded(data):
    a = str(data)
    if len(a) < 2:
        return a.zfill(2)
    else:
        return a

data["Class_Size"]["padded_csd"] = data["Class_Size"]["CSD"].apply(padded)
data["Class_Size"]["DBN"] = data["Class_Size"]["padded_csd"] + data["Class_Size"]["SCHOOL CODE"]
data["Class_Size"].head(3)

CSD BOROUGH SCHOOL CODE                SCHOOL NAME GRADE  PROGRAM TYPE  \
0    1       M        M015  P.S. 015 Roberto Clemente     0K       GEN ED   
1    1       M        M015  P.S. 015 Roberto Clemente     0K          CTT   
2    1       M        M015  P.S. 015 Roberto Clemente     01       GEN ED   

  CORE SUBJECT (MS CORE and 9-12 ONLY) CORE COURSE (MS CORE and 9-12 ONLY)  \
0                                    -                                   -   
1                                    -                                   -   
2                                    -                                   -   

  SERVICE CATEGORY(K-9* ONLY)  NUMBER OF STUDENTS / SEATS FILLED  \
0                           -                               19.0   
1                           -                               21.0   
2                           -                               17.0   

   NUMBER OF SECTIONS  AVERAGE CLASS SIZE  SIZE OF SMALLEST CLASS  \
0                 1.0                19.0                    19.0   
1                 1.0                21.0                    21.0   
2                 1.0                17.0                    17.0   

   SIZE OF LARGEST CLASS DATA SOURCE  SCHOOLWIDE PUPIL-TEACHER RATIO  \
0                   19.0         ATS                             NaN   
1                   21.0         ATS                             NaN   
2                   17.0         ATS                             NaN   

  padded_csd     DBN  
0         01  01M015  
1         01  01M015  
2         01  01M015

### Combining SAT Scores

Combining the exam's three SAT scores into one single number will make it much easier to correlate scores with demographic factors.

A new column will be created totaling the SAT scores after the *SAT Math Avg. Score*, *SAT Critical Reading Avg. Score*, and *SAT Writing Avg. Score* columns in the *Sat_Results* data set have been converted from the object (string) data type to a numeric data type.

In [11]:
cols = ['SAT Math Avg. Score', 'SAT Critical Reading Avg. Score', 'SAT Writing Avg. Score']

for c in cols:
    data['SAT_Results'][c] = pd.to_numeric(data['SAT_Results'][c], errors='coerce')

data['SAT_Results']['sat_score'] = data['SAT_Results'][cols[0]] + data['SAT_Results'][cols[1]] + data['SAT_Results'][cols[2]]
data['SAT_Results']['sat_score'].head()

0    1122.0
1    1172.0
2    1149.0
3    1174.0
4    1207.0
Name: sat_score, dtype: float64

### Parsing Geographic Coordinates for Schools

We will now parse the latitude and longitude coordinates for each school to map the schools and uncover any geographic patterns in the data. The coordinates are currently in the text field *Location 1* in the *HS_Directory* data set. The coordinates need to be extracted from the field for it contains a lot of unwanted information. 

In [12]:
import re

def latitude(loc):
    coord = re.findall("\(.+\)", loc)
    lat = coord[0].split(',')[0].replace("(","")
    return lat

def longitude(loc):
    coord = re.findall("\(.+\)", loc)
    lon = coord[0].split(',')[1].replace(")","")
    return lon

data['HS_Directory']['lat'] = data['HS_Directory']['Location 1'].apply(latitude)
data['HS_Directory']['lon'] = data['HS_Directory']['Location 1'].apply(longitude)

# Convert coordinates from strings to numbers
cols = ['lon','lat']
for c in cols:
    data['HS_Directory'][c] = pd.to_numeric(data['HS_Directory'][c], errors='coerce')

data['HS_Directory'].head(3)

dbn                                  school_name   borough  \
0  17K548          Brooklyn School for Music & Theatre  Brooklyn   
1  09X543             High School for Violin and Dance     Bronx   
2  09X327  Comprehensive Model School Project M.S. 327     Bronx   

  building_code  phone_number    fax_number  grade_span_min  grade_span_max  \
0          K440  718-230-6250  718-230-6262             9.0              12   
1          X400  718-842-0687  718-589-9849             9.0              12   
2          X240  718-294-8111  718-294-8109             6.0              12   

   expgrade_span_min  expgrade_span_max    ...      \
0                NaN                NaN    ...       
1                NaN                NaN    ...       
2                NaN                NaN    ...       

                                          Location 1 Community Board  \
0  883 Classon Avenue\nBrooklyn, NY 11225\n(40.67...             9.0   
1  1110 Boston Road\nBronx, NY 10456\n(40.8276026...             3.0   
2  1501 Jerome Avenue\nBronx, NY 10452\n(40.84241...             4.0   

  Council District Census Tract        BIN           BBL  \
0             35.0        213.0  3029686.0  3.011870e+09   
1             16.0        135.0  2004526.0  2.026340e+09   
2             14.0        209.0  2008336.0  2.028590e+09   

                                                 NTA     DBN        lat  \
0  Crown Heights South                           ...  17K548  40.670299   
1  Morrisania-Melrose                            ...  09X543  40.827603   
2  West Concourse                                ...  09X327  40.842414   

         lon  
0 -73.961648  
1 -73.904475  
2 -73.916162  

[3 rows x 67 columns]

# III. Combining the Data Sets
### Condensing Class Size Data

The main data set we want to analyze, *Sat_Results*, has unique DBN values for every high school in New York City while other data sets have duplicate DBN values. A single row in the *Sat_Results* data set may match multiple rows in the *Class_Size* data set, for example. This situation will create problems, because we don't know which of the multiple entries in the *Class_Size* data set we should combine with the single matching entry in sat_results. Before merging the data sets, we'll condense the class size, graduation, and demographics data sets so that each DBN is unique.

In [13]:
data['Class_Size'].head()

CSD BOROUGH SCHOOL CODE                SCHOOL NAME GRADE  PROGRAM TYPE  \
0    1       M        M015  P.S. 015 Roberto Clemente     0K       GEN ED   
1    1       M        M015  P.S. 015 Roberto Clemente     0K          CTT   
2    1       M        M015  P.S. 015 Roberto Clemente     01       GEN ED   
3    1       M        M015  P.S. 015 Roberto Clemente     01          CTT   
4    1       M        M015  P.S. 015 Roberto Clemente     02       GEN ED   

  CORE SUBJECT (MS CORE and 9-12 ONLY) CORE COURSE (MS CORE and 9-12 ONLY)  \
0                                    -                                   -   
1                                    -                                   -   
2                                    -                                   -   
3                                    -                                   -   
4                                    -                                   -   

  SERVICE CATEGORY(K-9* ONLY)  NUMBER OF STUDENTS / SEATS FILLED  \
0                           -                               19.0   
1                           -                               21.0   
2                           -                               17.0   
3                           -                               17.0   
4                           -                               15.0   

   NUMBER OF SECTIONS  AVERAGE CLASS SIZE  SIZE OF SMALLEST CLASS  \
0                 1.0                19.0                    19.0   
1                 1.0                21.0                    21.0   
2                 1.0                17.0                    17.0   
3                 1.0                17.0                    17.0   
4                 1.0                15.0                    15.0   

   SIZE OF LARGEST CLASS DATA SOURCE  SCHOOLWIDE PUPIL-TEACHER RATIO  \
0                   19.0         ATS                             NaN   
1                   21.0         ATS                             NaN   
2                   17.0         ATS                             NaN   
3                   17.0         ATS                             NaN   
4                   15.0         ATS                             NaN   

  padded_csd     DBN  
0         01  01M015  
1         01  01M015  
2         01  01M015  
3         01  01M015  
4         01  01M015

The first few rows all pertain to the same school, which is why the DBN appears more than once. Each school has multiple values for GRADE, PROGRAM TYPE, CORE SUBJECT (MS CORE and 9-12 ONLY), and CORE COURSE (MS CORE and 9-12 ONLY).

If we look at the unique values for GRADE, we get the following:

In [14]:
data["Class_Size"]["GRADE "].unique()

array(['0K', '01', '02', '03', '04', '05', '0K-09', nan, '06', '07', '08',
       'MS Core', '09-12', '09'], dtype=object)

Because we're dealing with high schools, we're only concerned with grades 9 through 12, i.e. where the value in the GRADE column is 09-12.

If we look at the unique values for PROGRAM TYPE, we get the following:

In [15]:
data["Class_Size"]["PROGRAM TYPE"].unique()

array(['GEN ED', 'CTT', 'SPEC ED', nan, 'G&T'], dtype=object)

Each school can have multiple program types. Because GEN ED is the largest category by far, let's only select rows where PROGRAM TYPE is GEN ED.

In [16]:
class_size = data["Class_Size"]
class_size = class_size[class_size["GRADE "] == "09-12"]
class_size = class_size[class_size["PROGRAM TYPE"] == "GEN ED"]
class_size.head(3)

CSD BOROUGH SCHOOL CODE                                    SCHOOL NAME  \
225    1       M        M292  Henry Street School for International Studies   
226    1       M        M292  Henry Street School for International Studies   
227    1       M        M292  Henry Street School for International Studies   

    GRADE  PROGRAM TYPE CORE SUBJECT (MS CORE and 9-12 ONLY)  \
225  09-12       GEN ED                              ENGLISH   
226  09-12       GEN ED                              ENGLISH   
227  09-12       GEN ED                              ENGLISH   

    CORE COURSE (MS CORE and 9-12 ONLY) SERVICE CATEGORY(K-9* ONLY)  \
225                           English 9                           -   
226                          English 10                           -   
227                          English 11                           -   

     NUMBER OF STUDENTS / SEATS FILLED  NUMBER OF SECTIONS  \
225                               63.0                 3.0   
226                               79.0                 3.0   
227                               38.0                 2.0   

     AVERAGE CLASS SIZE  SIZE OF SMALLEST CLASS  SIZE OF LARGEST CLASS  \
225                21.0                    19.0                   25.0   
226                26.3                    24.0                   31.0   
227                19.0                    16.0                   22.0   

    DATA SOURCE  SCHOOLWIDE PUPIL-TEACHER RATIO padded_csd     DBN  
225       STARS                             NaN         01  01M292  
226       STARS                             NaN         01  01M292  
227       STARS                             NaN         01  01M292

As we saw when we displayed class_size on the last screen, DBN still isn't completely unique. This is due to the CORE COURSE (MS CORE and 9-12 ONLY) and CORE SUBJECT (MS CORE and 9-12 ONLY) columns pertaining to different kinds of classes. For example, here are the unique values for CORE SUBJECT (MS CORE and 9-12 ONLY):

In [17]:
data["Class_Size"]["CORE SUBJECT (MS CORE and 9-12 ONLY)"].unique()

array(['-', nan, 'ENGLISH', 'MATH', 'SCIENCE', 'SOCIAL STUDIES'], dtype=object)

We want our class size data to include every single class a school offers, not just a subset of them. Taking the average across all of the classes a school offers will give us unique DBN values.

In [18]:
import numpy

class_size = class_size.groupby("DBN").agg(numpy.mean)
class_size.reset_index(inplace=True)
data["Class_Size"] = class_size
data["Class_Size"].head(3)

DBN  CSD  NUMBER OF STUDENTS / SEATS FILLED  NUMBER OF SECTIONS  \
0  01M292    1                               88.0                 4.0   
1  01M332    1                               46.0                 2.0   
2  01M378    1                               33.0                 1.0   

   AVERAGE CLASS SIZE  SIZE OF SMALLEST CLASS  SIZE OF LARGEST CLASS  \
0           22.564286                    18.5              26.571429   
1           22.000000                    21.0              23.500000   
2           33.000000                    33.0              33.000000   

   SCHOOLWIDE PUPIL-TEACHER RATIO  
0                             NaN  
1                             NaN  
2                             NaN

### Condensing Demographics Data

In [19]:
data['Demographics'].head(3)

DBN                       Name  schoolyear fl_percent  frl_percent  \
0  01M015  P.S. 015 ROBERTO CLEMENTE    20052006       89.4          NaN   
1  01M015  P.S. 015 ROBERTO CLEMENTE    20062007       89.4          NaN   
2  01M015  P.S. 015 ROBERTO CLEMENTE    20072008       89.4          NaN   

   total_enrollment prek   k grade1 grade2    ...     black_num black_per  \
0               281   15  36     40     33    ...            74      26.3   
1               243   15  29     39     38    ...            68      28.0   
2               261   18  43     39     36    ...            77      29.5   

  hispanic_num hispanic_per white_num white_per male_num male_per female_num  \
0          189         67.3         5       1.8    158.0     56.2      123.0   
1          153         63.0         4       1.6    140.0     57.6      103.0   
2          157         60.2         7       2.7    143.0     54.8      118.0   

  female_per  
0       43.8  
1       42.4  
2       45.2  

[3 rows x 38 columns]

The only column that prevents a given DBN from being unique is schoolyear. We only want to select rows where schoolyear is "20112012." This will give us the most recent year of data and match our SAT results data.

In [20]:
demographics = data["Demographics"]
demographics = demographics[demographics["schoolyear"] == 20112012]
data["Demographics"] = demographics
data["Demographics"].head(3)

DBN                                              Name  schoolyear  \
6   01M015  P.S. 015 ROBERTO CLEMENTE                           20112012   
13  01M019  P.S. 019 ASHER LEVY                                 20112012   
20  01M020  PS 020 ANNA SILVER                                  20112012   

   fl_percent  frl_percent  total_enrollment prek    k grade1 grade2  \
6         NaN         89.4               189   13   31     35     28   
13        NaN         61.5               328   32   46     52     54   
20        NaN         92.5               626   52  102    121     87   

      ...     black_num black_per hispanic_num hispanic_per white_num  \
6     ...            63      33.3          109         57.7         4   
13    ...            81      24.7          158         48.2        28   
20    ...            55       8.8          357         57.0        16   

   white_per male_num male_per female_num female_per  
6        2.1     97.0     51.3       92.0       48.7  
13       8.5    147.0     44.8      181.0       55.2  
20       2.6    330.0     52.7      296.0       47.3  

[3 rows x 38 columns]

### Condensing Graduation Data

In [21]:
data["Graduation"].head(3)

Demographic     DBN                            School Name Cohort  \
0  Total Cohort  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL   2003   
1  Total Cohort  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL   2004   
2  Total Cohort  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL   2005   

   Total Cohort Total Grads - n Total Grads - % of cohort Total Regents - n  \
0             5               s                         s                 s   
1            55              37                     67.3%                17   
2            64              43                     67.2%                27   

  Total Regents - % of cohort Total Regents - % of grads  \
0                           s                          s   
1                       30.9%                      45.9%   
2                       42.2%                      62.8%   

             ...            Regents w/o Advanced - n  \
0            ...                                   s   
1            ...                                  17   
2            ...                                  27   

  Regents w/o Advanced - % of cohort Regents w/o Advanced - % of grads  \
0                                  s                                 s   
1                              30.9%                             45.9%   
2                              42.2%                             62.8%   

  Local - n Local - % of cohort   Local - % of grads Still Enrolled - n  \
0         s                   s                    s                  s   
1        20               36.4%                54.1%                 15   
2        16                 25%  37.200000000000003%                  9   

  Still Enrolled - % of cohort Dropped Out - n Dropped Out - % of cohort  
0                            s               s                         s  
1                        27.3%               3                      5.5%  
2                        14.1%               9                     14.1%  

[3 rows x 23 columns]

The Demographic and Cohort columns prevent DBN from being unique. A Cohort refers to the year the data represents and the Demographic refers to a specific demographic group. We want data from the most recent Cohort available, which is 2006 and data from the full cohort, which is labeled "Total Cohort."

In [22]:
graduation = data["Graduation"]
graduation = graduation[graduation["Cohort"] == "2006"]
graduation = graduation[graduation["Demographic"] == "Total Cohort"]
data["Graduation"] = graduation
data["Graduation"].head(3)

Demographic     DBN                            School Name Cohort  \
3   Total Cohort  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL   2006   
10  Total Cohort  01M448    UNIVERSITY NEIGHBORHOOD HIGH SCHOOL   2006   
17  Total Cohort  01M450             EAST SIDE COMMUNITY SCHOOL   2006   

    Total Cohort Total Grads - n Total Grads - % of cohort Total Regents - n  \
3             78              43                     55.1%                36   
10           124              53                     42.7%                42   
17            90              70                     77.8%                67   

   Total Regents - % of cohort Total Regents - % of grads  \
3                        46.2%                      83.7%   
10                       33.9%                      79.2%   
17         74.400000000000006%                      95.7%   

              ...            Regents w/o Advanced - n  \
3             ...                                  36   
10            ...                                  34   
17            ...                                  67   

   Regents w/o Advanced - % of cohort Regents w/o Advanced - % of grads  \
3                               46.2%                             83.7%   
10                              27.4%                             64.2%   
17                74.400000000000006%                             95.7%   

   Local - n Local - % of cohort Local - % of grads Still Enrolled - n  \
3          7                  9%              16.3%                 16   
10        11                8.9%              20.8%                 46   
17         3                3.3%               4.3%                 15   

   Still Enrolled - % of cohort Dropped Out - n Dropped Out - % of cohort  
3                         20.5%              11                     14.1%  
10                        37.1%              20       16.100000000000001%  
17                        16.7%               5                      5.6%  

[3 rows x 23 columns]

### Converting AP Test Scores

In [23]:
data['AP_Results'].head(3)

DBN                    SchoolName AP Test Takers  Total Exams Taken  \
0  01M448  UNIVERSITY NEIGHBORHOOD H.S.              39                49   
1  01M450        EAST SIDE COMMUNITY HS              19                21   
2  01M515           LOWER EASTSIDE PREP              24                26   

  Number of Exams with scores 3 4 or 5  
0                                   10  
1                                    s  
2                                   24

AP exams have a 1 to 5 scale; 3 or higher is a passing score. AP exams are more common in academically challenging institutions than in schools that lack funding or academic rigor.

To find out whether AP exam scores are correlated with SAT scores across high schools, we need to convert the AP exam scores in the *AP_Results* data set from string to to numeric values.

Three columns need to be converted:
- AP Test Takers
- Total Exams Taken
- Number of Exams with scores 3 4 or 5

In [24]:
cols = ['AP Test Takers ', 'Total Exams Taken', 'Number of Exams with scores 3 4 or 5']
for i in cols:
    data["AP_Results"][i] = pd.to_numeric(data["AP_Results"][i], errors="coerce")

data["AP_Results"].dtypes

DBN                                      object
SchoolName                               object
AP Test Takers                          float64
Total Exams Taken                       float64
Number of Exams with scores 3 4 or 5    float64
dtype: object

### Merging the Data Sets

The DBN column will be used to identify matching rows across data sets. There may be DBN values that exist in one data set, but not in another. This is partly because the data is from different years. Each data set also has inconsistencies in terms of how it was gathered. Human error may also play a role. Therefore, we may not find matches for the DBN values in *Sat_Results* in all of the other data sets, and other data sets may have DBN values that don't exist in *Sat_Results*. 

Both the AP results and graduation data sets have many missing DBN values, so we'll use a "left" join when we merge the SAT results data set with them. Because we're using a "left" join, our final dataframe will have all of the same DBN values as the original *SAT_Results* dataframe.

In [25]:
combined = data["SAT_Results"]
combined = combined.merge(data["AP_Results"], how='left', on='DBN')
combined = combined.merge(data["Graduation"], how='left', on='DBN')
combined.head(3)

DBN                                    SCHOOL NAME  \
0  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL STUDIES   
1  01M448            UNIVERSITY NEIGHBORHOOD HIGH SCHOOL   
2  01M450                     EAST SIDE COMMUNITY SCHOOL   

  Num of SAT Test Takers  SAT Critical Reading Avg. Score  \
0                     29                            355.0   
1                     91                            383.0   
2                     70                            377.0   

   SAT Math Avg. Score  SAT Writing Avg. Score  sat_score  \
0                404.0                   363.0     1122.0   
1                423.0                   366.0     1172.0   
2                402.0                   370.0     1149.0   

                     SchoolName  AP Test Takers   Total Exams Taken  \
0                           NaN              NaN                NaN   
1  UNIVERSITY NEIGHBORHOOD H.S.             39.0               49.0   
2        EAST SIDE COMMUNITY HS             19.0               21.0   

             ...             Regents w/o Advanced - n  \
0            ...                                   36   
1            ...                                   34   
2            ...                                   67   

  Regents w/o Advanced - % of cohort Regents w/o Advanced - % of grads  \
0                              46.2%                             83.7%   
1                              27.4%                             64.2%   
2                74.400000000000006%                             95.7%   

  Local - n  Local - % of cohort Local - % of grads Still Enrolled - n  \
0         7                   9%              16.3%                 16   
1        11                 8.9%              20.8%                 46   
2         3                 3.3%               4.3%                 15   

  Still Enrolled - % of cohort Dropped Out - n Dropped Out - % of cohort  
0                        20.5%              11                     14.1%  
1                        37.1%              20       16.100000000000001%  
2                        16.7%               5                      5.6%  

[3 rows x 33 columns]

In [26]:
combined.shape

(479, 33)

Our new *combined* dataframe has a total of 479 rows and 33 columns.

Now we merge *Class_Size, Demographics, survey*, and *HS_Directory* into *combined*. Because these files contain information that's more valuable to our analysis and also have fewer missing DBN values, we'll use the "inner" join type.

In [27]:
datasets = ["Class_Size", "Demographics", "survey", "HS_Directory"]

for i in datasets:
    combined = combined.merge(data[i], how='inner', on='DBN')
    
combined.head()

DBN                                        SCHOOL NAME  \
0  01M292      HENRY STREET SCHOOL FOR INTERNATIONAL STUDIES   
1  01M448                UNIVERSITY NEIGHBORHOOD HIGH SCHOOL   
2  01M450                         EAST SIDE COMMUNITY SCHOOL   
3  01M509                            MARTA VALLE HIGH SCHOOL   
4  01M539  NEW EXPLORATIONS INTO SCIENCE, TECHNOLOGY AND ...   

  Num of SAT Test Takers  SAT Critical Reading Avg. Score  \
0                     29                            355.0   
1                     91                            383.0   
2                     70                            377.0   
3                     44                            390.0   
4                    159                            522.0   

   SAT Math Avg. Score  SAT Writing Avg. Score  sat_score  \
0                404.0                   363.0     1122.0   
1                423.0                   366.0     1172.0   
2                402.0                   370.0     1149.0   
3                433.0                   384.0     1207.0   
4                574.0                   525.0     1621.0   

                       SchoolName  AP Test Takers   Total Exams Taken  \
0                             NaN              NaN                NaN   
1    UNIVERSITY NEIGHBORHOOD H.S.             39.0               49.0   
2          EAST SIDE COMMUNITY HS             19.0               21.0   
3                             NaN              NaN                NaN   
4  NEW EXPLORATIONS SCI,TECH,MATH            255.0              377.0   

     ...      priority10                                         Location 1  \
0    ...             NaN  220 Henry Street\nNew York, NY 10002\n(40.7137...   
1    ...             NaN  200 Monroe Street\nNew York, NY 10002\n(40.712...   
2    ...             NaN  420 East 12 Street\nNew York, NY 10009\n(40.72...   
3    ...             NaN  145 Stanton Street\nNew York, NY 10002\n(40.72...   
4    ...             NaN  111 Columbia Street\nNew York, NY 10002\n(40.7...   

  Community Board Council District  Census Tract        BIN           BBL  \
0             3.0              1.0         201.0  1003223.0  1.002690e+09   
1             3.0              1.0         202.0  1003214.0  1.002590e+09   
2             3.0              2.0          34.0  1005974.0  1.004390e+09   
3             3.0              1.0        3001.0  1004323.0  1.003540e+09   
4             3.0              2.0        2201.0  1004070.0  1.003350e+09   

                                                 NTA        lat        lon  
0  Lower East Side                               ...  40.713764 -73.985260  
1  Lower East Side                               ...  40.712332 -73.984797  
2  East Village                                  ...  40.729783 -73.983041  
3  Chinatown                                     ...  40.720569 -73.985673  
4  Lower East Side                               ...  40.718725 -73.979426  

[5 rows x 165 columns]

In [28]:
combined.shape

(363, 165)

Our final *combined* dataframe has a total of 363 rows and 165 columns, 116 fewer rows than in *SAT_Results*. This is because pandas couldn't find the DBN values that existed in *SAT_Results* in the other data sets. While this is worth investigating, we're currently looking for high-level correlations, so we don't need to dive into which DBNs are missing.

There are now many columns with null (NaN) values. This is because we chose to do "left" joins, where some columns may not have had data. The data set also had some missing values to begin with. If we hadn't performed a "left" join, all of the rows with missing data would have been lost in the merge process, which wouldn't have left us with many high schools in our data set. 

We'll fill in the missing values with the means of their respective columns to avoid disrupting our analyses. If a column consists entirely of null or NaN values, pandas won't be able to fill in the missing values because there won't be a mean. Any remaining NaN or null values will be filled in with the value 0.

In [29]:
combined = combined.fillna(combined.mean())
combined = combined.fillna(0)
combined.head()

DBN                                        SCHOOL NAME  \
0  01M292      HENRY STREET SCHOOL FOR INTERNATIONAL STUDIES   
1  01M448                UNIVERSITY NEIGHBORHOOD HIGH SCHOOL   
2  01M450                         EAST SIDE COMMUNITY SCHOOL   
3  01M509                            MARTA VALLE HIGH SCHOOL   
4  01M539  NEW EXPLORATIONS INTO SCIENCE, TECHNOLOGY AND ...   

  Num of SAT Test Takers  SAT Critical Reading Avg. Score  \
0                     29                            355.0   
1                     91                            383.0   
2                     70                            377.0   
3                     44                            390.0   
4                    159                            522.0   

   SAT Math Avg. Score  SAT Writing Avg. Score  sat_score  \
0                404.0                   363.0     1122.0   
1                423.0                   366.0     1172.0   
2                402.0                   370.0     1149.0   
3                433.0                   384.0     1207.0   
4                574.0                   525.0     1621.0   

                       SchoolName  AP Test Takers   Total Exams Taken  \
0                               0       129.028846         197.038462   
1    UNIVERSITY NEIGHBORHOOD H.S.        39.000000          49.000000   
2          EAST SIDE COMMUNITY HS        19.000000          21.000000   
3                               0       129.028846         197.038462   
4  NEW EXPLORATIONS SCI,TECH,MATH       255.000000         377.000000   

     ...      priority10                                         Location 1  \
0    ...               0  220 Henry Street\nNew York, NY 10002\n(40.7137...   
1    ...               0  200 Monroe Street\nNew York, NY 10002\n(40.712...   
2    ...               0  420 East 12 Street\nNew York, NY 10009\n(40.72...   
3    ...               0  145 Stanton Street\nNew York, NY 10002\n(40.72...   
4    ...               0  111 Columbia Street\nNew York, NY 10002\n(40.7...   

  Community Board Council District  Census Tract        BIN           BBL  \
0             3.0              1.0         201.0  1003223.0  1.002690e+09   
1             3.0              1.0         202.0  1003214.0  1.002590e+09   
2             3.0              2.0          34.0  1005974.0  1.004390e+09   
3             3.0              1.0        3001.0  1004323.0  1.003540e+09   
4             3.0              2.0        2201.0  1004070.0  1.003350e+09   

                                                 NTA        lat        lon  
0  Lower East Side                               ...  40.713764 -73.985260  
1  Lower East Side                               ...  40.712332 -73.984797  
2  East Village                                  ...  40.729783 -73.983041  
3  Chinatown                                     ...  40.720569 -73.985673  
4  Lower East Side                               ...  40.718725 -73.979426  

[5 rows x 165 columns]

We now have a clean data set on which we can base our analysis. Mapping the statistics out on a school district level might be an interesting way to analyze them. Adding a column to the data set that specifies the school district will help us accomplish this.

The school district is just the first two characters of the DBN. We can apply a function over the DBN column of combined that pulls out the first two letters.

In [30]:
def district(DBN):
    return DBN[0:2]

combined["school_dist"] = combined["DBN"].apply(district)
combined["school_dist"].head(3)

0    01
1    01
2    01
Name: school_dist, dtype: object

[...]